docs(theming): generating color palettes at runtime (#2532)
IgniteUI/igniteui-docfx: en/components/themes/generating-color-palettes.md

---

> title: Generating color palettes at runtime

Reviewer: Theming with CSS Variables

---
title: Generating color palettes at runtime
_description: The Ignite UI for Angular theming engine provides an algorithm that generates Material-like palettes from base colors and contrast text colors for each hue in the palette.
_keywords: Ignite UI for Angular, UI controls, Angular widgets, web widgets, UI widgets, Angular, Native Angular Components Suite, Native Angular Controls, Native Angular Components Library, Native Angular Components, Angular Theming Component, Angular Theming
---

# Generating color palettes at runtime
<p class="highlight">The Ignite UI for Angular theming engine provides an algorithm that generates Material-like palettes from base colors and contrast text colors for each hue in the palette.</p>
<div class="divider"></div>

## Overview

> Our theming library has not only a global theme, that styles the entire application and every component in it, but also individual component themes where each component can use a different palette. Therefore, currently, we do not expose functionality for generating color palettes and switching the colors at runtime. The main obstacle to having this by default is that some of the colors generated for some of the components are not pure palette colors, instead, they are calculated at build time based on the component schema. They may start from a palette color, but then the color is scaled in one way or another using SASS color functions, which won't work with CSS variable references.

Reviewer: ```suggestion
Themes in Ignite UI for Angular are defined by _shape_, _shadows_, _typography_, and **color**. The default way of changing the colors in a theme is to pass a palette generated by the `igx-palette` function to the `igx-theme` mixin. This palette will contain over 40 color shades (generated at build time) that are all represented by HEX/rgba color values. These values will in turn be used by individual component themes to automatically generate contrasting colors at build time. For instance, the raised button uses the `secondary` color for its background. If the `secondary` color is set to a lighter hue, the raised button will automatically calculate its text color to a contrasting color that will adhere to the WCAG AAA accessibility requirements. This functionality truly shines in more complex components like the Grid or Calendar. This is great, however, there's one major drawback to this approach - you cannot pass references to colors not known at Sass build time, like custom CSS properties, to the `igx-palette` function.

Some of our users want to be able to change component theme colors at runtime and we allow that by exposing custom CSS variables for themes built with `$legacy-support` off. Obviously, this will not automatically generate contrasting colors as color values are not recalculated at runtime.

This article will expand on how users can pass custom CSS properties (CSS variables) to individual component themes so that colors can be 'connected' to a different palette, even external ones generated by other frameworks, like `Bootstrap` or `Angular Material`.
```

> However, using slight modifications and custom schemas or themes, we will be able to use any custom CSS property as a value.

Reviewer: ```suggestion
By modifying component schemas or themes, we will be able to use any custom CSS property as a value.
```

Let's take the [avatar]({environment:angularApiUrl}/classes/igxavatarcomponent.html) component for example.

> First, make sure we include a palette to some scope, in this case, the root scope is used:

Reviewer: ```suggestion
First, make sure to include the colors of a palette as CSS variables to some scope, in this case, the root scope is used:
```
